### ISSUE TITLE
Make ZM able to use Postgresql [$250]

### ISSUE THREAD

**Issue by connortechnology** [ZoneMinder/zoneminder#355]
I'm creating this as an issue so that people can post bounties to it.  

I've already done the php side by converting zm to PDO.  However the C needs to be changed and lots of db queries will need to have their syntax adjusted.  
##

**Comment by JoshuaPK** [comment]
What work is still left on this modification?  Also, as far as upgrading the schema: is it a requirement that the scripts do schema upgrades while the system is live, or would it be acceptable for some downtime?  Or, could we design the upgrade scripts such that one might do the following:

0. We create a slightly different (and incompatible) schema to be database agnostic;
1. Spin up a new ZoneMinder instance that is intended to replace a current ZM instance;
2. Script would pull the camera and other configuration information from the current instance to the new instance;
3. "Flip a switch" so that the cameras and monitoring are now running on the new ZM instance;
4. Script would pull all event and historical data from the old ZM instance and insert into the new ZM instance

I think if we could do a step-by-step version upgrade vs. an in-place upgrade, this modification would be easier.

**Comment by connortechnology** [comment]
I'm not sure anyone cares about migration. Not at this point at least.

What needs to be done is to change all the code and SQL calls to use some db-agnostic library.

**Comment by JoshuaPK** [comment]
Revisiting the abstraction layer again.  It looks like OpenDBX hasn't been updated for a while.  I've used another package called SOCI, which was updated as recently as two months ago.  What do you think about this?  https://github.com/SOCI/soci

**Comment by connortechnology** [comment]
Yeah SOCI looks to be about the right level for our needs. Is also packaged with ubuntu, so that is nice.

**Comment by zdzichu** [comment]
Closing this issue while it isn't solved is wrong. Please reopen.

**Comment by knight-of-ni** [comment]
Leaving issues open, that no human is interested in touching, is wrong. Thus, until someone steps up and convinces us they will do the work, this issue will remain closed.

**Comment by connortechnology** [comment]
I'm reopening because I am going to get around to it soon.  Going to make it a milestone for 1.35

**Comment by connortechnology** [comment]
Bountysource is going to take the money in this bounty on July 1st unless the funds are redirected.  Alternatively I can close this out, collect the bounty and ZoneMinder will hold the funds until it is properly closed.

**Comment by parvit** [comment]
Hi all,
i've been working on this for some time now and i think the direction is clear enough to discuss it.
My branch is at https://github.com/parvit/zoneminder/tree/issue-355
Beware it does not even compile all files with queries yet (up to before zm_monitor.cpp at the moment).

The changes are based on using the soci library, they contain *at the moment*:
- CMake soci library integration
- Implementation of the generic db interface and concrete one using mysql soci backend
- Implementation of the query executor and queue
- Porting of part of the zm classes to using the new interface

The approach however has been to map all queries to prepared statements initialized at the beginning by the selected concrete db backend and reusing it (binding the values in the spot instead of composing the query as text).
Also i've tried to minimize the syntax and usage of the query class to something manageable, let me know if it's clear enough or any change is necessary.

Some issues i'm working through:
- Soci mysql backend does not support nested queries
  - Will rework the 3 queries i've found to use joins instead
- Complex structures handled by the queries now need an adapter template class (```type_conversion<>```) which instructs soci how to handle a particular complex type in terms of the fields returned by a statement
  - For reasons of headers inclusion ordering, i've elected to put the type_conversion of a class in the header the type is declared (at first i tried a single header but it was not feasible without touching a lot more things in the process)

What is missing yet:
- [ ] Porting of the rest of zm classes to using the new interface (in progress)
- [ ] Implementation of concrete db interface with postgresql soci backend
- [ ] CMake support for the detection and availability of the db modules
- [ ] Configuration support for the selection of the requested backend

TLDR;
It's a lot of work ahead still.

**Comment by GreasyMonkee** [comment]
I am not a coder/developer, but am following this topic with a great deal of interest - am willing and happy to help by adding more to the bounty to help along the work on this.

**Comment by jlpoolen** [comment]
> I am not a coder/developer, but am following this topic with a great deal of interest - am willing and happy to help by adding more to the bounty to help along the work on this.

Take a look at project moonfire-nvr by Scott Lamb.

**Comment by connortechnology** [comment]
@jlpoolen This is not an acceptable comment.  There are lots of other places to discuss ZoneMinder alternatives.  Your comments add nothing relevant to the discussion here. 

@GreasyMonkee I'm sure @parvit would appreciate that.  This is a huge effort.

**Comment by GreasyMonkee** [comment]
Hi @connortechnology @parvit,

Some questions.

1. Reading through #3644, I see that you have the YAML for Centos8 there - I am using Fedora 37, I can try to re-write the Centos8 YAML to match for Fedora (finding the appropriate packages, etc)

2. In Fedora, the Postgresql packaging is a bit of a mess - with what appears to be a "Fedora-specific" package from the Fedora Repositories, and the "official" Postgresql PGDG release packages, with a slightly different naming, i.e. "postgresql" is "postgresql15", as per the "PostgreSQL Global Development Group" naming convention - is it possible to write the YAML in such a fashion that it looks initially for "postgresql" and if not found (repository or packaged disabled, etc), then try "postgresql && $releasever" for the installation?

3. Further to this, and I am showing my Newb status here - besides the packages listed in the YAML scripts, are there any other packages that are expected to be there? I am totally new to ZoneMinder, so please pardon my ignorance on this.....

**Comment by parvit** [comment]
@GreasyMonkee Do you have a specific version of fedora in mind? I could create a yaml specific for that distro.

**Comment by GreasyMonkee** [comment]
@parvit I would say Fedora 36, as this is where Postgresql14 was released, however now there is Postgresql15 for Fedora Rawhide (becoming Fedora 38) - I personally will run Fedora 37 on the ZM Server
If I remember correctly there was some significant changes before Fedora 36, so I would not try to go back further than that.

**Comment by parvit** [comment]
@GreasyMonkee Added the flow for fedora 36, works locally in docker but won't work (like the other flows) on the builders until the master is fixed too.
